components/ca-input-array.js

```diff
@@ -0,0 +1,631 @@
+define([
+    './helpers/create-element.js', 
+    './helpers/clear-element.js', 
+    './helpers/cancel-event.js', 
+    'document-register-element'
+], (createElement, clearElement, cancelEvent) => {
+
+    /**
+     * @exports ca-input-array component
+     * @extends HTMLElement
+     */
+    class CaInputArray extends HTMLElement {
+
+        /**
+         * Executes when the element is first created
+         * @access private
+         * @returns {void}
+         */
+        createdCallback() {
+
+            this.render();
+        }
+
+        /**
+         * Executes when any attribute is changed on the element
+         * @access private
+         * @type {Event}
+         * @param {string} attrName - the name of the attribute to have changed
+         * @param {string} oldVal - the old value of the attribute
+         * @param {string} newVal - the new value of the attribute
+         * @returns {void}
+         */
+        attributeChangedCallback(attrName, oldVal, newVal) {
+
+            // if we do not have a property matching the attribute, exit
+            if (!this[attrName]) return;
+
+            // it the value has changed, update and re-render to reflect changes
+            if (oldVal !== newVal) {
+                this[attrName] = newVal;
+                this.render();
+            }
+        }
+
+        /**
+         * @returns {string} type
+         */
+        get type() {
+            return this.getAttribute('type') || 'text';
+        }
+
+        /**
+         * @param {string} value type
+         * @returns {void}
+         */
+        set type(value) {
+            this.setAttribute('type', value);
+        }
+
+        /**
+         * @returns {string} id
+         */
+        get id() {
+            return this.getAttribute('id') || '';
+        }
+
+        /**
+         * @param {string} value id
+         * @returns {void}
+         */
+        set id(value) {
+            this.setAttribute('id', value);
+        }
+
+        /**
+         * @returns {array|string} value
+         */
+        get value() {
+
+            const self = this;
+            const elements = Array.prototype.slice(this.querySelectorAll(this.childSelector));
+            let values = [];
+            let invalid = false;
+
+            // grab value for each item
+            values = elements.map(item => item.value || '');
+
+            // remove empty items
+            values = values.filter(Boolean);
+
+            if (this.type === 'number') {
+
+                // convert values to numbers
+                values = values.map(value => parseInt(value || '0', 10));
+
+                // preform .min .max .divisibleBy validation (if configured)
+                values = values.filter(num => {
+
+                    const isValid = this.isValidNumber(self, num);
```

Comment:
Arrow function, meaning that `this` is passed down. self shouldn't be required as it's safe to pass this.